src/awkward1/operations/convert.py

```diff
@@ -1139,6 +1139,180 @@ def getfunction(layout, depth):
+                return pyarrow.StructArray.from_arrays(values, layout.keys())
+        
+        elif isinstance(layout, (awkward1.layout.UnionArray8_32, 
+                                awkward1.layout.UnionArray8_64, 
+                                awkward1.layout.UnionArray8_U32)):
+        
+            values = [recurse(x) for x in layout.contents]
+            if mask is not None:
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)
+                return pyarrow.UnionArray.from_dense(recurse(layout.tags), recurse(layout.index), values).take(pyarrow.array(indices))
+            return pyarrow.UnionArray.from_dense(recurse(layout.tags), recurse(layout.index), values)
+        
+        elif isinstance(layout, (awkward1.layout.IndexedArray32, 
+                                awkward1.layout.IndexedArray64, 
+                                awkward1.layout.IndexedArrayU32)):
+            if mask is not None:
+                indices = []
+                for i in range(len(mask)):
+                    if mask[i] == 1:
+                        indices.append(None)
+                    else:
+                        indices.append(i)
+                return pyarrow.DictionaryArray.from_arrays(recurse(layout.index), recurse(layout.content)).take(pyarrow.array(indices))
+            else:
+                return pyarrow.DictionaryArray.from_arrays(recurse(layout.index), recurse(layout.content))
+
+        elif isinstance(layout, (awkward1.layout.IndexedOptionArray32, 
+                                awkward1.layout.IndexedOptionArray64)):
+            # Convert this to a Union Array
+            dummy_numeric_val = awkward1.Array([]).layout
+            tags = numpy.zeros(len(layout.index), dtype=numpy.int8)
+            index = numpy.asarray(layout.index)
+            for i in range(len(index)):
+                if index[i] < 0:
+                    index[i] = 0
+                    tags[i] = 0
+                else:
+                    tags[i] = 1
+            array = awkward1.layout.UnionArray8_32(awkward1.layout.Index8(tags), awkward1.layout.Index32(index), [dummy_numeric_val, layout.content])
+            return(recurse(array))
+        
+        elif isinstance(layout, (awkward1.layout.BitMaskedArray)):
+            mask = numpy.asarray(layout.bytemask(), dtype=numpy.bool) == (
+                not layout.valid_when)
+            # Keep the other elements as null if len(mask) < len(layout.content)
+            # mask = numpy.append(mask, [True for x in range(len(layout.content) - len(mask))])
+            return recurse(layout.content, mask).slice(0, len(mask))
+
+        elif isinstance(layout, (awkward1.layout.ByteMaskedArray)):
+            mask = numpy.asarray(layout.mask, dtype=numpy.bool) == (not layout.valid_when)
```

Comment:
Please let me know if this is the correct way to parse the mask.

Reply:
Each of the option-type nodes (IndexedOptionArray, ByteMaskedArray, BitMaskedArray, UnmaskedArray) has a `bytemask()` method for making masks that NumPy's MaskedArray likes; they should also have a `bitmask()` method for making masks that Arrow likes (favorable choice of `valid_when` and `lsb_order`). That could be done in NumPy with `packbits` (and a trick to work around the fact that NumPy picked the wrong `lsb_order`) or it could be a new C++ method (`bitmask()`).